queries/nix.scm

```diff
@@ -0,0 +1,312 @@
+)
+
+;;
+;; Conditionals
+;;
+
+;;!! key = if a then b else c;
+;;!        ^^^^^^^^^^^^^^^^^^
+(if_expression) @ifStatement @branch.iteration @condition.iteration
+;;!! key = if a then b else c;
+;;!        ^^^^^^^^^^^
+;;!        xxxxxxxxxxxx
+(if_expression
+    "if" @branch.start
+    condition: (_)
+    "then"
+    consequence: (_) @branch.end
+)
+;;!! key = if a then b else c;
+;;!                    ^^^^^^
+;;!                   xxxxxxx
+(if_expression
+    "else" @branch.start
+    alternative: (_) @branch.end
+)
+
+;;!! key = if a > 10 then b else c;
+;;!           ^^^^^^
+(if_expression
+    condition: (_) @condition
+) @_.domain
+
+;;
+;; Lists and maps
+;;
+
+;;!! foo = [ a b c ];
+;;!          ^
+;;!          xx
+(list_expression
+    element: (_) @collectionItem
+)
+
+;;!! foo = [ a b c ];
+;;!        ^^^^^^^^^
+(list_expression
+    element: (_)+ @list.interior
+) @list
+
+;;!! foo = { x = 1; y = 2; };
+;;!        ^^^^^^^^^^^^^^^^^
+[
+    (attrset_expression
+        (_) @map.interior
+    )
+    (rec_attrset_expression
+        (_) @map.interior
+    )
+] @map @statement.iteration @value.iteration @name.iteration
+
+;;!! foo = { x = 1; y = 2; };
+;;!          ^
+;;!          xxxx
+;;!          -----
+(binding
+    attrpath: (_) @collectionKey @_.trailing.start.endOf
+    expression: (_) @_.trailing.end.startOf
+) @_.domain
+
+;;
+;; Strings
+;;
+
+;;!! # foo
+;;!  ^^^^^
+(comment) @comment @textFragment
+
+[
+    (string_expression)
+    (indented_string_expression)
+] @string
+(string_fragment) @textFragment
+
+;;!! d = "foo ${bar} baz"
+;;!           ^^^^^^
+;;!      <**************>
+(string_expression
+    (interpolation) @argumentOrParameter
+) @_.iteration
+
+;;
+;; Functions
+;;
+
+;; Note for this part of the function, we identify is as lambda only
+;;!! x = a: b: a + b;
+;;!         ^^^^^^^^ Func1 due to currying
+;;!      ^^^^^^^^^^^ Func2 due to currying
+(function_expression
+    [
+        (
+            ;; Match foo@{ a, b }: style
+            (identifier) @argumentOrParameter.start
+            "@"
+            (formals) @argumentOrParameter.end
+        )
+        ;; Match a: style arg
+        (identifier) @argumentOrParameter
+
+        ;; Match { a, b }: style
+        (formals) @argumentOrParameter
+    ]
+    .
+    body: (_) @anonymousFunction.interior
+) @_.domain @anonymousFunction @argumentOrParameter.iteration
+
+;; We define the named function as the full assignment of the lambda
+;; This means funk name becomes the variable holding the lambda
+(binding
+    (_) @functionName
+    "="
+    expression: (function_expression) @namedFunction.interior
+) @namedFunction @functionName.domain
+
+;; Calls to functions are a bit finicky, because of everything being curried lambdas
+(
+    (apply_expression) @dummy @functionCall @argumentOrParameter.iteration
+    (#not-parent-type? @functionCall apply_expression)
+)
+
+;; This is gross, but not sure how to do fuzzy matching against an unknown number of
```

Comment:
I'm curious if someone is aware of a better way to do this? I couldn't see anything in other .scm files, the tree sitter query documentation, or chatgpt. Assuming there isn't something I missed, it almost seems like something a special predicate could handle more elegantly?